types/src/circom.rs

```diff
@@ -0,0 +1,210 @@
+    ] {
+        gamma_abc_g1.push(points);
+    }
+
+    let vk = VerifyingKey {
+        alpha_g1,
+        beta_g2,
+        gamma_g2,
+        delta_g2,
+        gamma_abc_g1,
+    };
+
+    PreparedVerifyingKey::from(vk)
+}
+
+fn parse_field_element(s: &str) -> Result<Fq, CryptoMaterialError> {
+    s.parse::<Fq>()
+        .map_err(|_e| CryptoMaterialError::DeserializationError)
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Hash, Serialize)]
+pub struct G1Projective {
+    x: String,
+    y: String,
+    z: String,
+}
+
+impl G1Projective {
+    pub fn new(x: &str, y: &str) -> Self {
+        G1Projective {
+            x: x.to_owned(),
+            y: y.to_owned(),
+            z: "1".to_string(),
+        }
+    }
+
+    pub fn to_affine(&self) -> Result<G1Affine, CryptoMaterialError> {
+        self.try_into()
+    }
+}
+
+impl TryInto<G1Affine> for &G1Projective {
+    type Error = CryptoMaterialError;
+
+    fn try_into(self) -> Result<G1Affine, CryptoMaterialError> {
+        let g1 = ark_bn254::G1Projective::new_unchecked(
+            parse_field_element(&self.x)?,
+            parse_field_element(&self.y)?,
+            parse_field_element(&self.z)?,
+        );
+        Ok(g1.into())
+    }
+}
+
+pub type Fq2Str = [String; 2];
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Hash, Serialize)]
+pub struct G2Projective {
+    x: Fq2Str,
+    y: Fq2Str,
+    z: Fq2Str,
+}
+
+impl G2Projective {
+    pub fn new(x: [&str; 2], y: [&str; 2]) -> Self {
+        G2Projective {
+            x: [x[0].to_owned(), x[1].to_owned()],
+            y: [y[0].to_owned(), y[1].to_owned()],
+            z: ["1".to_string(), "0".to_owned()],
+        }
+    }
+
+    pub fn to_affine(&self) -> Result<G2Affine, CryptoMaterialError> {
+        self.try_into()
+    }
+}
+
+impl TryInto<G2Affine> for &G2Projective {
+    type Error = CryptoMaterialError;
+
+    fn try_into(self) -> Result<G2Affine, CryptoMaterialError> {
+        let g2 = ark_bn254::G2Projective::new_unchecked(
+            Fq2::new(
+                parse_field_element(&self.x[0])?,
+                parse_field_element(&self.x[1])?,
+            ),
+            Fq2::new(
+                parse_field_element(&self.y[0])?,
+                parse_field_element(&self.y[1])?,
+            ),
+            Fq2::new(
+                parse_field_element(&self.z[0])?,
+                parse_field_element(&self.z[1])?,
+            ),
+        );
+        Ok(g2.into())
+    }
+}
+
+pub fn get_public_inputs_hash(
+    sig: &ZkIdSignature,
+    pk: &ZkIdPublicKey,
+    jwk: &RSA_JWK,
+) -> anyhow::Result<ark_bn254::Fr> {
+    // Add the epk as padded and packed scalars
+    let mut frs = poseidon_bn254::pad_and_pack_bytes_to_scalars_with_len(
+        sig.ephemeral_pubkey.to_bytes().as_slice(),
+        MAX_EPK_BYTES,
+    )?;
+
+    // Add the id_commitment as a scalar
+    frs.push(ark_bn254::Fr::from_le_bytes_mod_order(&pk.idc.0));
+
+    let iat_val = 1700255944;
```

Comment:
We cannot merge until this is taken care of right?

Reply:
Done